dist/tools/bmp/bmp.py

```diff
@@ -0,0 +1,236 @@
+parser.add_argument('--swd', action='store_true', help='use SWD transport (default)')
+parser.add_argument('--connect-srst', action='store_true', help='reset target while connecting')
+parser.add_argument('--tpwr', action='store_true', help='enable target power')
+parser.add_argument('--serial', help='choose specific probe by serial number')
+parser.add_argument('--port', help='choose specific probe by port')
+parser.add_argument('--attach', help='choose specific target by number', default='1')
+parser.add_argument('--gdb-path', help='path to GDB', default='gdb-multiarch')
+parser.add_argument('--term-cmd', help='serial terminal command',
+                    default='picocom --nolock --imap lfcrlf --baud 115200 %s')
+parser.add_argument('action', help='choose a task to perform', nargs='?',
+                    choices=['list', 'flash', 'erase', 'debug', 'term', 'reset'],
+                    default='list')
+parser.add_argument('file', help='file to load to target (hex or elf)', nargs='?')
+
+TIMEOUT = 100  # seconds
+
+
+# find all connected BMPs and store both GDB and UART interfaces
+def detect_probes():
+    GDBs = []
+    UARTs = []
+    for p in serial.tools.list_ports.comports():
+        if p.vid == 0x1D50 and p.pid in {0x6018, 0x6017}:
+            if re.fullmatch(r'COM\d\d', p.device):
+                p.device = '//./' + p.device
+            if 'GDB' in str(p.interface) \
+                    or re.fullmatch(r'/dev/cu\.usbmodem([A-F0-9]*)1', p.device) \
+                    or p.location[-1] == '0' and os.name == 'nt':
+                GDBs.append(p)
+            else:
+                UARTs.append(p)
+    return GDBs, UARTs
+
+
+# search device with specific serial number <snr> in list <l>
+def search_serial(snr, l):
+    for p in l:
+        if snr in p.serial_number:
+            return p.device
+
+
+# parse GDB output for targets
+def detect_targets(gdbmi, res):
+    targets = []
+    while True:
+        for msg in res:
+            if msg['type'] == 'target':
+                m = re.fullmatch(pattern=r"\s*(\d)+\s*(.*)\\n", string=msg['payload'])
+                if m:
+                    targets.append(m.group(2))
+            elif msg['type'] == 'result':
+                assert msg['message'] == 'done', str(msg)
+                return targets
+
+        res = gdbmi.get_gdb_response(timeout_sec=TIMEOUT)
+
+
+def gdb_write_and_wait_for_result(gdbmi, cmd, description, expected_result='done'):
+    res = gdbmi.write(cmd, timeout_sec=TIMEOUT)
+    while True:
+        for msg in res:
+            if msg['type'] == 'result':
+                if msg['message'] == expected_result:
+                    print(description, "successful.")
+                    return True
+                else:
+                    print(description, "failed.", file=sys.stderr)
+                    return False
+        res = gdbmi.get_gdb_response(timeout_sec=TIMEOUT)
+
+
+if __name__ == '__main__':
```

Comment:
At least some of the code below could go into its own function for easier readability.

Reply:
Do you mean code from the main function?

Reply:
Well. there is no main function ;-). But yes, moving some of the stuff below into their own functions definitely helped with readability.

Reply:
Yeah, you're right. ^^' What do think about the code? Is it readable now?

Reply:
@miri64: Ping?

Reply:
Better :-)